// Code generated by protoc-gen-gogo. DO NOT EDIT.
// source: app.proto

package main

import (
	"context"
	fmt "fmt"
	"github.com/coreos/etcd/clientv3/concurrency"
	_ "github.com/gogo/googleapis/google/api"
	_ "github.com/gogo/protobuf/gogoproto"
	proto "github.com/gogo/protobuf/proto"
	_ "github.com/mobiledgex/edge-cloud/d-match-engine/dme-proto"
	"github.com/mobiledgex/edge-cloud/edgeproto"
	"github.com/mobiledgex/edge-cloud/objstore"
	_ "github.com/mobiledgex/edge-cloud/protogen"
	"github.com/stretchr/testify/require"
	math "math"
	"testing"
)

// Reference imports to suppress errors if they are not otherwise used.
var _ = proto.Marshal
var _ = fmt.Errorf
var _ = math.Inf

// Auto-generated code: DO NOT EDIT

// Caller must write by hand the test data generator.
// Each Ref object should only have a single reference to the key,
// in order to properly test each reference (i.e. don't have a single
// object that has multiple references).
type AppDeleteDataGen interface {
	GetAppTestObj() (*edgeproto.App, *testSupportData)
	GetTrustPolicyExceptionKeyAppKeyRef(key *edgeproto.AppKey) (*edgeproto.TrustPolicyException, *testSupportData)
	GetAppAppInstInstsRef(key *edgeproto.AppKey) (*edgeproto.AppInstRefs, *testSupportData)
}

// AppDeleteStore wraps around the usual
// store to instrument checks and inject data while
// the delete api code is running.
type AppDeleteStore struct {
	edgeproto.AppStore
	t                   *testing.T
	allApis             *AllApis
	putDeletePrepare    bool
	putDeletePrepareCb  func()
	putDeletePrepareSTM concurrency.STM
}

func (s *AppDeleteStore) Put(ctx context.Context, m *edgeproto.App, wait func(int64), ops ...objstore.KVOp) (*edgeproto.Result, error) {
	if wait != nil {
		s.putDeletePrepare = m.DeletePrepare
	}
	res, err := s.AppStore.Put(ctx, m, wait, ops...)
	if s.putDeletePrepare && s.putDeletePrepareCb != nil {
		s.putDeletePrepareCb()
	}
	return res, err
}

func (s *AppDeleteStore) STMPut(stm concurrency.STM, obj *edgeproto.App, ops ...objstore.KVOp) {
	// there's an assumption that this is run within an ApplySTMWait,
	// where we wait for the caches to be updated with the transaction.
	if obj.DeletePrepare {
		s.putDeletePrepare = true
		s.putDeletePrepareSTM = stm
	} else {
		s.putDeletePrepare = false
		s.putDeletePrepareSTM = nil
	}
	s.AppStore.STMPut(stm, obj, ops...)
	if s.putDeletePrepare && s.putDeletePrepareCb != nil {
		s.putDeletePrepareCb()
	}
}

func (s *AppDeleteStore) Delete(ctx context.Context, m *edgeproto.App, wait func(int64)) (*edgeproto.Result, error) {
	require.True(s.t, s.putDeletePrepare, "DeletePrepare must be comitted to database with a sync.Wait before deleting")
	return s.AppStore.Delete(ctx, m, wait)
}

func (s *AppDeleteStore) STMDel(stm concurrency.STM, key *edgeproto.AppKey) {
	require.True(s.t, s.putDeletePrepare, "DeletePrepare must be comitted to database with a sync.Wait before deleting")
	s.AppStore.STMDel(stm, key)
}

func (s *AppDeleteStore) requireUndoDeletePrepare(ctx context.Context, obj *edgeproto.App) {
	deletePrepare := s.getDeletePrepare(ctx, obj)
	require.False(s.t, deletePrepare, "must undo delete prepare field on failure")
}

func (s *AppDeleteStore) getDeletePrepare(ctx context.Context, obj *edgeproto.App) bool {
	buf := edgeproto.App{}
	found := s.Get(ctx, obj.GetKey(), &buf)
	require.True(s.t, found, "expected test object to be found")
	return buf.DeletePrepare
}

func deleteAppChecks(t *testing.T, ctx context.Context, all *AllApis, dataGen AppDeleteDataGen) {
	var err error
	// override store so we can inject data and check data
	api := all.appApi
	origStore := api.store
	deleteStore := &AppDeleteStore{
		AppStore: origStore,
		t:        t,
		allApis:  all,
	}
	api.store = deleteStore
	origappInstRefsApiStore := all.appInstRefsApi.store
	appInstRefsApiStore := &AppInstRefsDeleteStore{
		AppInstRefsStore: origappInstRefsApiStore,
	}
	all.appInstRefsApi.store = appInstRefsApiStore
	defer func() {
		api.store = origStore
		all.appInstRefsApi.store = origappInstRefsApiStore
	}()

	// inject testObj directly, bypassing create checks/deps
	testObj, supportData := dataGen.GetAppTestObj()
	supportData.put(t, ctx, all)
	defer supportData.delete(t, ctx, all)
	origStore.Put(ctx, testObj, api.sync.syncWait)

	// Positive test, delete should succeed without any references.
	// The overrided store checks that delete prepare was set on the
	// object in the database before actually doing the delete.
	// This call back checks that any refs lookups are done in the
	// same stm as the delete prepare is set.
	deleteStore.putDeletePrepareCb = func() {
		// make sure ref objects reads happen in same stm
		// as delete prepare is set
		require.NotNil(t, deleteStore.putDeletePrepareSTM, "must set delete prepare in STM")
		require.NotNil(t, appInstRefsApiStore.getRefSTM, "must check for refs from AppInstRefs in STM")
		require.Equal(t, deleteStore.putDeletePrepareSTM, appInstRefsApiStore.getRefSTM, "delete prepare and ref check for AppInstRefs must be done in the same STM")
	}
	testObj, _ = dataGen.GetAppTestObj()
	_, err = api.DeleteApp(ctx, testObj)
	require.Nil(t, err, "delete must succeed with no refs")
	deleteStore.putDeletePrepareCb = nil

	// Negative test, inject testObj with delete prepare already set.
	testObj, _ = dataGen.GetAppTestObj()
	testObj.DeletePrepare = true
	origStore.Put(ctx, testObj, api.sync.syncWait)
	// delete should fail with already being deleted
	testObj, _ = dataGen.GetAppTestObj()
	_, err = api.DeleteApp(ctx, testObj)
	require.NotNil(t, err, "delete must fail if already being deleted")
	require.Contains(t, err.Error(), "already being deleted")
	// failed delete must not interfere with existing delete prepare state
	require.True(t, deleteStore.getDeletePrepare(ctx, testObj), "delete prepare must not be modified by failed delete")

	// inject testObj for ref tests
	testObj, _ = dataGen.GetAppTestObj()
	origStore.Put(ctx, testObj, api.sync.syncWait)

	{
		// Negative test, TrustPolicyException refers to App.
		// The cb will inject refBy obj after delete prepare has been set.
		refBy, supportData := dataGen.GetTrustPolicyExceptionKeyAppKeyRef(testObj.GetKey())
		supportData.put(t, ctx, all)
		deleteStore.putDeletePrepareCb = func() {
			all.trustPolicyExceptionApi.store.Put(ctx, refBy, all.trustPolicyExceptionApi.sync.syncWait)
		}
		testObj, _ = dataGen.GetAppTestObj()
		_, err = api.DeleteApp(ctx, testObj)
		require.NotNil(t, err, "must fail delete with ref from TrustPolicyException")
		require.Contains(t, err.Error(), "in use")
		// check that delete prepare was reset
		deleteStore.requireUndoDeletePrepare(ctx, testObj)
		// remove TrustPolicyException obj
		_, err = all.trustPolicyExceptionApi.store.Delete(ctx, refBy, all.trustPolicyExceptionApi.sync.syncWait)
		require.Nil(t, err, "cleanup ref from TrustPolicyException must succeed")
		deleteStore.putDeletePrepareCb = nil
		supportData.delete(t, ctx, all)
	}
	{
		// Negative test, AppInstRefs refers to App via refs object.
		// Inject the refs object to trigger an "in use" error.
		refBy, supportData := dataGen.GetAppAppInstInstsRef(testObj.GetKey())
		supportData.put(t, ctx, all)
		_, err = all.appInstRefsApi.store.Put(ctx, refBy, all.appInstRefsApi.sync.syncWait)
		require.Nil(t, err)
		testObj, _ = dataGen.GetAppTestObj()
		_, err = api.DeleteApp(ctx, testObj)
		require.NotNil(t, err, "delete with ref from AppInstRefs must fail")
		require.Contains(t, err.Error(), "in use")
		// check that delete prepare was reset
		deleteStore.requireUndoDeletePrepare(ctx, testObj)
		// remove AppInstRefs obj
		_, err = all.appInstRefsApi.store.Delete(ctx, refBy, all.appInstRefsApi.sync.syncWait)
		require.Nil(t, err, "cleanup ref from AppInstRefs must succeed")
		supportData.delete(t, ctx, all)
	}

	// clean up testObj
	testObj, _ = dataGen.GetAppTestObj()
	_, err = api.DeleteApp(ctx, testObj)
	require.Nil(t, err, "cleanup must succeed")
}
